staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/matcher.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generic
+
+import (
+	celtypes "github.com/google/cel-go/common/types"
+
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/plugin/cel"
+	"k8s.io/klog/v2"
+)
+
+// matcher implements the Matcher interface
+type matcher struct {
+	filter     cel.Filter
+}
+
+func NewMatcher(filter cel.Filter) Matcher {
+	return &matcher{
+		filter:     filter,
+	}
+}
+
+// Match takes a list of Evaluation converts them into actionable a decision of whether they match or not.
+// If no match returns name of failed condition
+func (v *matcher) Match(versionedAttr *admission.VersionedAttributes) (bool, string, error) {
+
+	evalResults, err := v.filter.ForInput(versionedAttr, nil, cel.CreateAdmissionRequest(versionedAttr.Attributes))
+	if err != nil {
+		// on error ignore the match condition, don't apply matching
```

Comment:
Sure, added a todo about an error event

Reply:
Should that code eventually live here (which is shared with embedded admission) or in webhook.go?

Reply:
fail open vs. fail closed on error here ins interesting.

Fail closed would mean sending the request to the webhook for admission? If the webhook is not designed to handle request that the matchCondition was suppose to filter out, then that could cause problems.

Fail open would mean skipping the webhook? For webhooks that are suppose to enforce something critical, and have `FailurePolicy: Fail`, this might be surprising?

Should `FailurePolicy` be extended control how errors are handled here?

Reply:
So what's interesting is that the failPolicy here is different than the failPolicy for ValidatingAdmissionPolicy in a way. For VAP the validation is the CEL evaluations so failurePolicy handling errors there makes sense. When I think of failurePolicy for VWH it's a failure calling the webhook. This is an error deciding if it should even call the webhook so imo it's a little different, that's why initially we just went with ignoring the errors. I'm open to doing it either way though

Reply:
I agree. `failPolicy` for `ValidatingAdmissionPolicy` more easily and directly can be applied to `matchConditions`.

I do think we need a better plan than "ignore if error" here however as that does not give the webhook author any way to prevent unwanted fail open semantics when using a webhook as a defense mechanism.

Any thoughts @tallclair @maxsmythe @liggitt ?

Reply:
So it wouldn't fail open in the sense that things would avoid the webhook, what will happen here is that additional things would be included and sent to the webhook as the matchConditions can only be used to reduce the scope of a webhook.

If we go the other way for example (or made it configurable), a bad evaluation of a rule would basically make the webhook useless as it wouldn't target anything at all because we're doing an AND of the matchConditions.

Reply:
With type checking, I'm assuming that errors here should be rare. I'm guessing the most likely cause of an error would be using secondary authz, and a failure calling an authorization webhook? In that case, I'm guessing that would be used for skipping the webhook if a user had a certain privilege. In that case, the safe thing would be to call the webhook anyway.

For the case of the webhook being called on a resource it doesn't support - that should really be handled by the rules, not the MatchConditions. In fact, we may want to add a comment that rules & namespace selector should be used over MatchConditions when possible.

Reply:
Will type checking always be effective? Thinking of CRDs and uses of embedded resources here.

There is also the possibility of non-schema errors, like divide-by-zero.

In my mind, the CEL code here is essentially an extension of the webhook code itself and using FailurePolicy makes sense:

If failure policy is `Fail`, then I read that as the user saying "if I'm not sure what the webhook's response to this would have been, we should be pessimistic", `Ignore` would have the opposite interpretation.

One benefit of the current "on failure, call the webhook anyway" approach is that on the off chance the webhook should have been called, the system will behave appropriately, maybe allowing resources that should be allowed that would otherwise be rejected.

One downside is that for "ignore on failure" webhooks, you may get over-enforcement: the exclude logic that someone wrote is no longer properly excluding things.

The comparison seems like a wash to me, so I'd go with whatever would be least surprising to a user. IMO that would be to use failure policy, but I could certainly see using docs or similar to set a different expectation.

TL;DR either works, but I have a slight preference for @jpbetz 's approach.

Reply:
So one thing though is if the user wants the webhook to be pessimistic and block things when its unsure, using the same failurePolicy for matchConditions would have the opposite effect. For example, if the failPolicy causes the match condition to evaluate to false, then the match condition will not match and the webhook will not get called and the request will be allowed in.

Reply:
Failure policy == fail doesn't mean match condition error => no match, it would mean match condition error => reject the request